File: apple-ds-core/01-数组与链表/linked_list.py
"""
单链表和双链表实现
"""

class ListNode:
    """链表节点"""
    
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    def __str__(self):
        return str(self.val)


class LinkedList:
    """单链表实现"""
    
    def __init__(self):
        """初始化"""
        self._head = None
        self._size = 0
    
    def __len__(self):
        """返回链表长度"""
        return self._size
    
    def is_empty(self):
        """判断是否为空"""
        return self._size == 0
    
    def prepend(self, val):
        """在头部添加节点
        
        Args:
            val: 节点值
        """
        new_node = ListNode(val, self._head)
        self._head = new_node
        self._size += 1
    
    def append(self, val):
        """在尾部添加节点
        
        Args:
            val: 节点值
        """
        if self._head is None:
            self._head = ListNode(val)
        else:
            curr = self._head
            while curr.next:
                curr = curr.next
            curr.next = ListNode(val)
        self._size += 1
    
    def insert(self, index, val):
        """在指定位置插入节点
        
        Args:
            index: 插入位置
            val: 节点值
            
        Raises:
            IndexError: 索引越界
        """
        if not 0 <= index <= self._size:
            raise IndexError('Index out of range')
        
        if index == 0:
            self.prepend(val)
            return
        
        curr = self._head
        for _ in range(index - 1):
            curr = curr.next
        
        curr.next = ListNode(val, curr.next)
        self._size += 1
    
    def remove(self, val):
        """删除第一个匹配的节点
        
        Args:
            val: 要删除的值
            
        Raises:
            ValueError: 值不存在
        """
        if self._head is None:
            raise ValueError(f'{val} not in list')
        
        # 删除头节点
        if self._head.val == val:
            self._head = self._head.next
            self._size -= 1
            return
        
        # 删除其他节点
        curr = self._head
        while curr.next:
            if curr.next.val == val:
                curr.next = curr.next.next
                self._size -= 1
                return
            curr = curr.next
        
        raise ValueError(f'{val} not in list')
    
    def pop(self, index=None):
        """删除并返回指定位置的节点值
        
        Args:
            index: 索引，默认为尾部
            
        Returns:
            节点值
            
        Raises:
            IndexError: 索引越界或链表为空
        """
        if self._head is None:
            raise IndexError('pop from empty list')
        
        if index is None:
            index = self._size - 1
        
        if not 0 <= index < self._size:
            raise IndexError('Index out of range')
        
        # 删除头节点
        if index == 0:
            val = self._head.val
            self._head = self._head.next
            self._size -= 1
            return val
        
        # 删除其他节点
        curr = self._head
        for _ in range(index - 1):
            curr = curr.next
        
        val = curr.next.val
        curr.next = curr.next.next
        self._size -= 1
        return val
    
    def find(self, val):
        """查找值的索引
        
        Args:
            val: 要查找的值
            
        Returns:
            索引
            
        Raises:
            ValueError: 值不存在
        """
        curr = self._head
        index = 0
        while curr:
            if curr.val == val:
                return index
            curr = curr.next
            index += 1
        raise ValueError(f'{val} not in list')
    
    def reverse(self):
        """反转链表"""
        prev = None
        curr = self._head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        self._head = prev
    
    def to_list(self):
        """转换为Python列表"""
        result = []
        curr = self._head
        while curr:
            result.append(curr.val)
            curr = curr.next
        return result
    
    def __str__(self):
        """字符串表示"""
        return ' -> '.join(str(val) for val in self.to_list()) + ' -> None'


# 双链表节点
class DListNode:
    """双链表节点"""
    
    def __init__(self, val=0, prev=None, next=None):
        self.val = val
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    """双链表实现"""
    
    def __init__(self):
        """初始化"""
        self._head = None
        self._tail = None
        self._size = 0
    
    def __len__(self):
        """返回链表长度"""
        return self._size
    
    def is_empty(self):
        """判断是否为空"""
        return self._size == 0
    
    def prepend(self, val):
        """在头部添加节点"""
        new_node = DListNode(val, None, self._head)
        if self._head is None:
            self._head = self._tail = new_node
        else:
            self._head.prev = new_node
            self._head = new_node
        self._size += 1
    
    def append(self, val):
        """在尾部添加节点"""
        new_node = DListNode(val, self._tail, None)
        if self._tail is None:
            self._head = self._tail = new_node
        else:
            self._tail.next = new_node
            self._tail = new_node
        self._size += 1
    
    def pop_front(self):
        """删除并返回头节点值"""
        if self._head is None:
            raise IndexError('pop from empty list')
        
        val = self._head.val
        self._head = self._head.next
        if self._head:
            self._head.prev = None
        else:
            self._tail = None
        self._size -= 1
        return val
    
    def pop_back(self):
        """删除并返回尾节点值"""
        if self._tail is None:
            raise IndexError('pop from empty list')
        
        val = self._tail.val
        self._tail = self._tail.prev
        if self._tail:
            self._tail.next = None
        else:
            self._head = None
        self._size -= 1
        return val
    
    def to_list(self):
        """转换为Python列表"""
        result = []
        curr = self._head
        while curr:
            result.append(curr.val)
            curr = curr.next
        return result
    
    def __str__(self):
        """字符串表示"""
        return ' <-> '.join(str(val) for val in self.to_list())


def demo():
    """演示链表的使用"""
    print("=== 单链表演示 ===\n")
    
    # 创建链表
    ll = LinkedList()
    print(f"创建空链表: {ll}\n")
    
    # 添加元素
    print("尾部添加: 1, 2, 3")
    for i in range(1, 4):
        ll.append(i)
    print(f"链表: {ll}\n")
    
    # 头部添加
    print("头部添加: 0")
    ll.prepend(0)
    print(f"链表: {ll}\n")
    
    # 插入
    print("在索引2插入99")
    ll.insert(2, 99)
    print(f"链表: {ll}\n")
    
    # 查找
    print(f"查找99的索引: {ll.find(99)}\n")
    
    # 删除
    print("删除元素99")
    ll.remove(99)
    print(f"链表: {ll}\n")
    
    # 反转
    print("反转链表")
    ll.reverse()
    print(f"链表: {ll}\n")
    
    print("\n=== 双链表演示 ===\n")
    
    # 创建双链表
    dll = DoublyLinkedList()
    print(f"创建空双链表\n")
    
    # 添加元素
    print("尾部添加: 1, 2, 3")
    for i in range(1, 4):
        dll.append(i)
    print(f"双链表: {dll}\n")
    
    # 头部添加
    print("头部添加: 0")
    dll.prepend(0)
    print(f"双链表: {dll}\n")
    
    # 头部删除
    print(f"头部删除: {dll.pop_front()}")
    print(f"双链表: {dll}\n")
    
    # 尾部删除
    print(f"尾部删除: {dll.pop_back()}")
    print(f"双链表: {dll}")


if __name__ == '__main__':
    demo()

File: web-sites-hub/frontend-docs/apple-ds-core-docs/01-array-and-linked-list/linked_list.cpp
/**
 * 单链表和双链表实现
 */

#include <iostream>
#include <stdexcept>

// 单链表节点
template<typename T>
struct ListNode {
    T val;
    ListNode* next;
    
    ListNode(T v) : val(v), next(nullptr) {}
};

// 单链表
template<typename T>
class LinkedList {
private:
    ListNode<T>* head;
    int size;

public:
    LinkedList() : head(nullptr), size(0) {}
    
    ~LinkedList() {
        clear();
    }
    
    int getSize() const {
        return size;
    }
    
    bool isEmpty() const {
        return size == 0;
    }
    
    // 头部添加
    void prepend(const T& val) {
        ListNode<T>* newNode = new ListNode<T>(val);
        newNode->next = head;
        head = newNode;
        size++;
    }
    
    // 尾部添加
    void append(const T& val) {
        if (head == nullptr) {
            head = new ListNode<T>(val);
        } else {
            ListNode<T>* curr = head;
            while (curr->next) {
                curr = curr->next;
            }
            curr->next = new ListNode<T>(val);
        }
        size++;
    }
    
    // 插入
    void insert(int index, const T& val) {
        if (index < 0 || index > size) {
            throw std::out_of_range("Index out of range");
        }
        
        if (index == 0) {
            prepend(val);
            return;
        }
        
        ListNode<T>* curr = head;
        for (int i = 0; i < index - 1; i++) {
            curr = curr->next;
        }
        
        ListNode<T>* newNode = new ListNode<T>(val);
        newNode->next = curr->next;
        curr->next = newNode;
        size++;
    }
    
    // 删除
    void remove(const T& val) {
        if (head == nullptr) {
            throw std::runtime_error("Remove from empty list");
        }
        
        // 删除头节点
        if (head->val == val) {
            ListNode<T>* temp = head;
            head = head->next;
            delete temp;
            size--;
            return;
        }
        
        // 删除其他节点
        ListNode<T>* curr = head;
        while (curr->next) {
            if (curr->next->val == val) {
                ListNode<T>* temp = curr->next;
                curr->next = curr->next->next;
                delete temp;
                size--;
                return;
            }
            curr = curr->next;
        }
        
        throw std::runtime_error("Value not found");
    }
    
    // 反转
    void reverse() {
        ListNode<T>* prev = nullptr;
        ListNode<T>* curr = head;
        while (curr) {
            ListNode<T>* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        head = prev;
    }
    
    // 查找
    int find(const T& val) const {
        ListNode<T>* curr = head;
        int index = 0;
        while (curr) {
            if (curr->val == val) {
                return index;
            }
            curr = curr->next;
            index++;
        }
        return -1;
    }
    
    // 清空
    void clear() {
        while (head) {
            ListNode<T>* temp = head;
            head = head->next;
            delete temp;
        }
        size = 0;
    }
    
    // 打印
    void print() const {
        ListNode<T>* curr = head;
        while (curr) {
            std::cout << curr->val;
            if (curr->next) std::cout << " -> ";
            curr = curr->next;
        }
        std::cout << " -> NULL" << std::endl;
    }
};


// 双链表节点
template<typename T>
struct DListNode {
    T val;
    DListNode* prev;
    DListNode* next;
    
    DListNode(T v) : val(v), prev(nullptr), next(nullptr) {}
};

// 双链表
template<typename T>
class DoublyLinkedList {
private:
    DListNode<T>* head;
    DListNode<T>* tail;
    int size;

public:
    DoublyLinkedList() : head(nullptr), tail(nullptr), size(0) {}
    
    ~DoublyLinkedList() {
        clear();
    }
    
    int getSize() const {
        return size;
    }
    
    bool isEmpty() const {
        return size == 0;
    }
    
    // 头部添加
    void prepend(const T& val) {
        DListNode<T>* newNode = new DListNode<T>(val);
        if (head == nullptr) {
            head = tail = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
        }
        size++;
    }
    
    // 尾部添加
    void append(const T& val) {
        DListNode<T>* newNode = new DListNode<T>(val);
        if (tail == nullptr) {
            head = tail = newNode;
        } else {
            newNode->prev = tail;
            tail->next = newNode;
            tail = newNode;
        }
        size++;
    }
    
    // 头部删除
    T popFront() {
        if (head == nullptr) {
            throw std::runtime_error("Pop from empty list");
        }
        
        T val = head->val;
        DListNode<T>* temp = head;
        head = head->next;
        if (head) {
            head->prev = nullptr;
        } else {
            tail = nullptr;
        }
        delete temp;
        size--;
        return val;
    }
    
    // 尾部删除
    T popBack() {
        if (tail == nullptr) {
            throw std::runtime_error("Pop from empty list");
        }
        
        T val = tail->val;
        DListNode<T>* temp = tail;
        tail = tail->prev;
        if (tail) {
            tail->next = nullptr;
        } else {
            head = nullptr;
        }
        delete temp;
        size--;
        return val;
    }
    
    // 清空
    void clear() {
        while (head) {
            DListNode<T>* temp = head;
            head = head->next;
            delete temp;
        }
        tail = nullptr;
        size = 0;
    }
    
    // 打印
    void print() const {
        DListNode<T>* curr = head;
        while (curr) {
            std::cout << curr->val;
            if (curr->next) std::cout << " <-> ";
            curr = curr->next;
        }
        std::cout << std::endl;
    }
};


// 演示
int main() {
    std::cout << "=== 单链表演示 ===" << std::endl << std::endl;
    
    LinkedList<int> ll;
    
    std::cout << "尾部添加: 1, 2, 3" << std::endl;
    for (int i = 1; i <= 3; i++) {
        ll.append(i);
    }
    std::cout << "链表: ";
    ll.print();
    std::cout << std::endl;
    
    std::cout << "头部添加: 0" << std::endl;
    ll.prepend(0);
    std::cout << "链表: ";
    ll.print();
    std::cout << std::endl;
    
    std::cout << "在索引2插入99" << std::endl;
    ll.insert(2, 99);
    std::cout << "链表: ";
    ll.print();
    std::cout << std::endl;
    
    std::cout << "反转链表" << std::endl;
    ll.reverse();
    std::cout << "链表: ";
    ll.print();
    std::cout << std::endl;
    
    std::cout << "=== 双链表演示 ===" << std::endl << std::endl;
    
    DoublyLinkedList<int> dll;
    
    std::cout << "尾部添加: 1, 2, 3" << std::endl;
    for (int i = 1; i <= 3; i++) {
        dll.append(i);
    }
    std::cout << "双链表: ";
    dll.print();
    std::cout << std::endl;
    
    std::cout << "头部添加: 0" << std::endl;
    dll.prepend(0);
    std::cout << "双链表: ";
    dll.print();
    std::cout << std::endl;
    
    std::cout << "头部删除: " << dll.popFront() << std::endl;
    std::cout << "双链表: ";
    dll.print();
    std::cout << std::endl;
    
    std::cout << "尾部删除: " << dll.popBack() << std::endl;
    std::cout << "双链表: ";
    dll.print();
    
    return 0;
}

